File: src/server/api/modelParams.test.ts
import { describe, it, expect } from 'vitest';
import { validateModelParams } from './modelParams.js';

describe('validateModelParams', () => {
  it('accepts all fields absent', () => {
    expect(validateModelParams({}).ok).toBe(true);
  });

  it.each([
    { model: 'sonnet' as const },
    { model: 'opus' as const },
    { model: 'haiku' as const },
    { effort: 'low' as const },
    { effort: 'medium' as const },
    { effort: 'high' as const },
    { effort: 'max' as const },
    { subagent_model: 'sonnet' as const },
    { model: 'opus' as const, effort: 'xhigh' as const, subagent_model: 'haiku' as const },
    { model: 'sonnet' as const, effort: 'max' as const },
    { effort: 'high' as const },
  ])('accepts valid combination %o', (input) => {
    const r = validateModelParams(input);
    expect(r.ok).toBe(true);
  });

  it('rejects unknown model', () => {
    const r = validateModelParams({ model: 'turbo' as unknown as 'sonnet' });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.message).toMatch(/model/);
  });

  it('rejects unknown effort', () => {
    const r = validateModelParams({ effort: 'enormous' as unknown as 'max' });
    expect(r.ok).toBe(false);
  });

  it('rejects unknown subagent_model', () => {
    const r = validateModelParams({ subagent_model: 'mini' as unknown as 'haiku' });
    expect(r.ok).toBe(false);
  });

  it('rejects effort + haiku', () => {
    const r = validateModelParams({ model: 'haiku', effort: 'high' });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.message).toMatch(/haiku/i);
  });

  it('rejects xhigh + sonnet', () => {
    const r = validateModelParams({ model: 'sonnet', effort: 'xhigh' as 'max' });
    expect(r.ok).toBe(false);
    if (!r.ok) expect(r.message).toMatch(/xhigh/i);
  });

  it('rejects xhigh + haiku (haiku rule wins first)', () => {
    const r = validateModelParams({ model: 'haiku', effort: 'xhigh' as 'max' });
    expect(r.ok).toBe(false);
  });

  it('accepts xhigh + opus', () => {
    const r = validateModelParams({ model: 'opus', effort: 'xhigh' });
    expect(r.ok).toBe(true);
  });

  it('accepts effort without model (model absent)', () => {
    // Server-side is permissive; UI prevents odd combos, server ignores unset model.
    const r = validateModelParams({ effort: 'high' });
    expect(r.ok).toBe(true);
  });

  it('treats null values identically to undefined', () => {
    expect(validateModelParams({ model: null, effort: null, subagent_model: null }).ok).toBe(true);
    expect(validateModelParams({ model: 'opus', effort: null, subagent_model: null }).ok).toBe(true);
  });
});

File: server-elixir/test/fbi/runs/model_params_test.exs
defmodule FBI.Runs.ModelParamsTest do
  use ExUnit.Case, async: true
  alias FBI.Runs.ModelParams

  test "accepts unset values" do
    assert :ok = ModelParams.validate(%{})
    assert :ok = ModelParams.validate(%{model: nil, effort: nil, subagent_model: nil})
  end

  test "accepts known model + effort combos" do
    assert :ok = ModelParams.validate(%{model: "sonnet", effort: "high"})
    assert :ok = ModelParams.validate(%{model: "opus", effort: "xhigh"})
    assert :ok = ModelParams.validate(%{model: "haiku"})
    assert :ok = ModelParams.validate(%{subagent_model: "sonnet"})
  end

  test "rejects unknown model" do
    assert {:error, "invalid model: gpt"} = ModelParams.validate(%{model: "gpt"})
  end

  test "rejects unknown effort" do
    assert {:error, "invalid effort: blast"} = ModelParams.validate(%{effort: "blast"})
  end

  test "rejects unknown subagent_model" do
    assert {:error, "invalid subagent_model: gpt"} =
             ModelParams.validate(%{subagent_model: "gpt"})
  end

  test "rejects effort on haiku" do
    assert {:error, "effort is not supported on haiku"} =
             ModelParams.validate(%{model: "haiku", effort: "low"})
  end

  test "rejects xhigh on non-opus" do
    assert {:error, "xhigh effort is only supported on opus"} =
             ModelParams.validate(%{model: "sonnet", effort: "xhigh"})
  end
end
